### PR TITLE
async: Async callback typedef and signal helper

### PR DESCRIPTION
**Pull request by teburd**
A simple common callback typedef and helper to connect it to a
k_poll_signal for moving spi/i2c async calls towards callbacks.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

### CODE REVIEW

File: include/zephyr/sys/async.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+ *
+ * May also be implemented by an application with the understanding
+ * that the callback is most likely to be done in the ISR context
+ * with the limitations and restrictions that imposes.
+ *
+ * @param result The result of the asynchronous call
+ * @param userdata Generic userdata requested to be tied
+ *        to the callback.
+ */
+typedef void (*async_callback_t)(int result, void *userdata);
```
Comment (henrikbrixandersen) [inline include/zephyr/sys/async.h:35]: As discussed earlier in the API meeting, I think the common async callback function signature needs to be able to take a context. For device drivers, the context will be the `const struct device *` for the device associated with the notification.

```suggestion
typedef void (*async_callback_t)(void *context, int result, void *userdata);
```

File: include/zephyr/sys/async.h [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+/**
+ * Asynchronous semaphore completion callback
+ *
+ * Sets the result and gives the semaphore
+ *
+ * @param result Result to assign to the async_sem
+ * @param userdata Pointer to a struct async_sem
+ */
+void async_semaphore_cb(int result, void *userdata);
```
Comment (henrikbrixandersen) [inline include/zephyr/sys/async.h:58]: I see the idea behind using `userdata` as the argument name, but it can be a bit confusing - this really isn't "user data" as the user is not free to pass arbitrary data pointers here (which goes against all the existing places we have a `void *user_data` argument in callbacks) - it must be a `struct async_sem *`.

I suggest naming the parameter to reflect the data type to be passed. This goes for all the callbacks.

Comment (henrikbrixandersen) [reply include/zephyr/sys/async.h:58]: I think each of these helper functions needs to state which context they can be called from (kernel, isr, userspace, ..?). Right now, I think all support being called from isr context, but this may change down the road when more helpers are added.

File: include/zephyr/sys/async.h (hunk 1 above)
Comment (henrikbrixandersen) [reply include/zephyr/sys/async.h:35]: Perhaps the `struct async_*`  structures could include a `void *context` pointer as well?

File: include/zephyr/sys/async.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+ * that the callback is most likely to be done in the ISR context
+ * with the limitations and restrictions that imposes.
+ *
+ * @param result The result of the asynchronous call
+ * @param userdata Generic userdata requested to be tied
+ *        to the callback.
+ */
+typedef void (*async_callback_t)(int result, void *userdata);
+
+#ifdef CONFIG_ASYNC
```
Comment (henrikbrixandersen) [inline include/zephyr/sys/async.h:37]: Should the API really be conditional? The implementation, I can understand (even if it wont result in anything but compile-time overhead if unconditional) - but the API?

File: include/zephyr/sys/async.h [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+/**
+ * Asynchronous work completion callback
+ *
+ * Submits an associated k_work task to the global work queue
+ *
+ * @param result Result to assign to async_work
+ * @param userdata Pointer to struct async_work
+ */
+void async_work_cb(int result, void *userdata);
```
Comment (henrikbrixandersen) [inline include/zephyr/sys/async.h:110]: Another useful helper would be for submitting work to a specific queue. The queue could be specified as a field in `struct async_work`. It could even be generalized as to always take a queue in `struct async_work` and use the global work queue if `NULL`.

**Review by henrikbrixandersen — DISMISSED** [review]
@teburd This looks really useful, thank you. A few comments below:

File: include/zephyr/sys/async.h (hunk 1 above)
Comment (teburd) [reply include/zephyr/sys/async.h:35]: Where would that be used? An example would be good to help me understand the use case

Comment (henrikbrixandersen) [reply include/zephyr/sys/async.h]: It would be used by the consumer of the async event, similar as to where `result` would be used.

File: include/zephyr/sys/async.h [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+ * Asynchronous semaphore completion callback
+ *
+ * Sets the result and gives the semaphore
+ *
+ * @note May be called from an ISR
+ *
+ * @param context Optional context from which the callback originated,
+ *                commonly a struct device*.
+ * @param result Result to assign to the async_sem
+ * @param sem_data Pointer to a struct async_sem
```
Comment (nordic-krch) [inline include/zephyr/sys/async.h:65]: no sure if doxygen resolves it but maybe better to add `@ref async_sem`

**Review by nordic-krch — COMMENTED** [review]
Looks ok, still we need a decision if we replace `notify` and how to do it. I will try to propose changes in `onoff` that will use `async` (but won't happen this week).

**Review by andyross — DISMISSED** [review]
Drive-by +1 while I try to avoid the API bikeshed.

Broadly: this looks perfectly fine and very reasonable, but the danger with any "framework" feature is that it ends up half-implemented.  I certainly won't stand in the way of this merging but I guess my gut says we should wait until we get as much of the existing APIs ported to it as possible (with some kind of compatibility layer to manage the deprecation period).

**Review by dcpleung — DISMISSED** [review]

**Review by tbursztyka — DISMISSED** [review]
lgtm

**Review by carlescufi — CHANGES_REQUESTED** [review]
I think we should deprecate/remove `sys_notify` as part of this PR, or remove it altogether?

**Comment by carlescufi** [comment]
> Looks ok, still we need a decision if we replace `notify` and how to do it. I will try to propose changes in `onoff` that will use `async` (but won't happen this week).

I suggest we do this as part of this PR, or in an immediate follow-up. @nordic-krch could you post a commit somewhere showing how `onoff` could use `async` instead of `sys_notify`?

File: lib/os/async.c [hunk 6]
```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright (c) 2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "toolchain/gcc.h"
```
Comment (gmarull) [inline lib/os/async.c:7]: why is this needed? why not `<zephyr/toolchain.h>`?

File: lib/os/async.c [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+void async_work_cb(void *context, int result, void *work_data)
+{
+	ARG_UNUSED(context);
+
+	struct async_work *a = (struct async_work *)work_data;
+
+	a->result = result;
+
+	if (a->workq) {
```
Comment (gmarull) [inline lib/os/async.c:52]:
```suggestion
	if (a->workq != NULL) {
```

File: include/zephyr/sys/async.h [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+	struct k_work_q *workq;
+};
+
+/**
+ * Asynchronous work completion callback
+ *
+ * Submits an associated k_work task to a work queue, the global
+ * work queue if async_work has a NULL workq.
+ *
+ * @note May be called from an ISR
```
Comment (gmarull) [inline include/zephyr/sys/async.h:128]:
```suggestion
 * @funcprops \isr_ok
```

File: include/zephyr/sys/async.h [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+#ifndef ZEPHYR_INCLUDE_SYS_ASYNC_H_
+#define ZEPHYR_INCLUDE_SYS_ASYNC_H_
+
+#include <zephyr/kernel.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
```
Comment (gmarull) [inline include/zephyr/sys/async.h:27]: note that content in this header doesn't seem to belong to any group, I think it should and at the same time be pulled in Sphinx side (with some usage examples maybe, like many Kernel APIs do)

**Review by gmarull — COMMENTED** [review]
lgtm, a few nitpicks

File: lib/os/async.c [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+void async_semaphore_cb(void *context, int result, void *sem_data)
+{
+	ARG_UNUSED(context);
+
+	struct async_sem *a = (struct async_sem *)sem_data;
+
+	a->result = result;
+
+	k_sem_give(a->sem);
```
Comment (carlescufi) [inline lib/os/async.c:19]: Note: `context` here is not propagated to the thread blocking with `k_sem_take()`

**Comment by carlescufi** [comment]
API meeting:

Outstanding issues:

1. `typedef void (*async_callback_t)(void *context, int result, void *data);`: Should this be `const void *context` instead so that this signature matches the standard driver callbacks that have `const struct device *dev` as the first parameter?
2. Direct propagation of `context` to the users of the async framework, to avoid users having to do that independently themselves

**Comment by teburd** [comment]
> API meeting:
> 
> Outstanding issues:
> 
>     1. `typedef void (*async_callback_t)(void *context, int result, void *data);`: Should this be `const void *context` instead so that this signature matches the standard driver callbacks that have `const struct device *dev` as the first parameter?

Happy to change the parameter to whatever folks wish it to be. 
 
>     2. Direct propagation of `context` to the users of the async framework, to avoid users having to do that independently themselves

I'm not grasping what the use case of this is to be honest. I understood that perhaps other async_callback_t implementations may want an additional context parameter signaling the callback originator. That makes sense! If the idea is to connect multiple async device calls to something like a k_poll_signal or k_sem or k_mutex... that seems like a misuse or potential race e.g.

``` c
struct async_sem asem = {
    .sem = &some_sem;
};

spi_transfer_async(spi_dev, ..., &asem);
i2c_transfer_async(i2c_dev, ..., &asem);
```

What is asem.context when the semaphore is taken? It's a dice roll I'd think?

Why not provide it and set it when creating the asem struct? e.g.

``` c
struct async_sem asem = {
  .sem = &some_sem;
  .context = &some_context;
};
```

Because a CONTAINER_OF can do just as good of a job if an authors puts the k_sem/k_mutex/k_work somewhere, with the added bonus its consistent and works well even with k_work and is the way described in the docs (which I'm happy to document here as well).



 

For a further argument on this front, k_poll_signal provides no direct context, it provides an int result and presumably it is not being connected to multiple async requests (contexts). The intended primary use case here is to replace k_poll_signal with a callback and some helpers in async apis (specifically spi and i2c).

In fact the async_signal_cb currently expects a struct k_poll_signal * as its data. The benefit here is that while changing _async APIs to use callbacks and userdata is a breaking change, it is an easy one to refactor today with perhaps a simple sed command adding the additional async_signal_cb param right before the existing k_poll_signal * param when calling them. Perhaps a sed friendly refactor even.

> 
>     3. Deprecation/removal of `sys_notify`, pending a draft commit by @nordic-krch

From my perspective this is entirely optional but I get the idea here that they do similar things, though for different purposes. Less (duplication) is always more of course.

**Comment by teburd** [comment]
Closing this was unintentional

**Comment by nordic-krch** [comment]
I've done initial work to remove `notify` use (#46192). There is one driver left which is a bit more difficult.
However, there is one thing I noticed. Maybe it was mentioned at API call. By introducing a callback with signature `cb(void *context, ...)` we kind of implicitly assume what context will be. For example, if we have a call like
```
foo_tx(const struct device *dev, uint8_t *data, size_t len, async_callback_t *cb, void *user_data);
```
we can only assume (or verbally inform in doxygen) that `context` will be the `dev` device. There is no C API way to commit to this agreement. `Notify` was doing it by having generic callback `cb()` and a module specific function which converted to a module specific callback. It was much more complex but cleaner.

**Comment by henrikbrixandersen** [comment]
> I'm not grasping what the use case of this is to be honest. I understood that perhaps other async_callback_t implementations may want an additional context parameter signaling the callback originator. That makes sense! If the idea is to connect multiple async device calls to something like a k_poll_signal or k_sem or k_mutex... that seems like a misuse or potential race e.g.
> 
> ```c
> struct async_sem asem = {
>     .sem = &some_sem;
> };
> 
> spi_transfer_async(spi_dev, ..., &asem);
> i2c_transfer_async(i2c_dev, ..., &asem);
> ```
> 
> What is asem.context when the semaphore is taken? It's a dice roll I'd think?

Of course (same goes for the `result` in the same struct). 

> Why not provide it and set it when creating the asem struct? e.g.
> 
> ```c
> struct async_sem asem = {
>   .sem = &some_sem;
>   .context = &some_context;
> };
> ```
> 
> Because a CONTAINER_OF can do just as good of a job if an author puts the k_sem/k_mutex/k_work in a container struct to hold the added info. For k_work this is the documented method and I'm happy to document that here as well.

The intention of providing the context within the async struct is not to lose data in the callback chain. Sure, each application can store this struct within another struct, fill it in, call CONTAINER_OF when receiving the notification etc., but the whole point of introducing this feature is to standardize on one pattern and reduce application level boilerplate code. Having a standardized way of retrieving the context from a callback will help further reduce the amount of application level code needed to retrieve the same context.

If the context is not provided, and the application needs it, it would be simpler to not use the async helpers at all and just implement your own callback (which gives access to both context and result) and trigger the IPC from there than having to put structures withing structures just to get to what was thrown away in the async helper callback chain.

File: include/zephyr/sys/async.h [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+ * Semaphore completion object with return code.
+ */
+struct async_sem {
+	/** Result of async call */
+	int result;
+
+	/** Semaphore that will be given on completion */
+	struct k_sem *sem;
+};
```
Comment (henrikbrixandersen) [inline include/zephyr/sys/async.h:56]: It could further reduce application level boilerplate code if we added initializer macros for all the `struct async_*` types.

Comment (teburd) [reply include/zephyr/sys/async.h]: Added _INITIALIZER macros

File: lib/os/async.c [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+#include <zephyr/sys/async.h>
+
+void async_semaphore_cb(void *context, int result, void *sem_data)
+{
+	ARG_UNUSED(context);
+
+	struct async_sem *a = (struct async_sem *)sem_data;
+
+	a->result = result;
+
```
Comment (carlescufi) [inline lib/os/async.c:17]: `a->context = context`

**Comment by carlescufi** [comment]
API meeting:

- @henrikbrixandersen to describe the usage of the context pointer with code
- @nordic-krch describe the usage of the context pointer in sys_notify with code
- @teburd to assess those and come back with a proposal

**Comment by gregshue** [comment]
Based on my past experience with product line platform code, here are a couple things to keep in mind:

a. Not all end users will have the freedom to use a modified Zephyr repository.  Some are only allowed to reuse what is provided.
b. When defining APIs, assume some users will need to have multiple instances of any arbitrary component.
c. When defining APIs, assume that downstream users will need to extend "enumerations" (e.g., result code defined values) as well as structures (e.g., via CONTAINER_OF())

The above points lead to needing both context and user data parameters. Lets consider when two identical instances exist of communication hardware (e.g., I2C controllers, ethernet controllers, USB OTG controllers, etc).  We will create a separate driver instance for each and may create a separate protocol stack instance for each.  The protocol component instance data is accessible via `userdata`, the driver component instance is provided through `context`.  If the protocol instance is only talking to one driver instance, then `context` may not be needed, but if it is talking to multiple instances of the same component (e.g., timers, GPIOs, etc.), then `context` is used to determine the source of the callback instead of providing a separate callback function for each source.

**Comment by teburd** [comment]
> Based on my past experience with product line platform code, here are a couple things to keep in mind:
> 
> a. Not all end users will have the freedom to use a modified Zephyr repository. Some are only allowed to reuse what is provided. b. When defining APIs, assume some users will need to have multiple instances of any arbitrary component. c. When defining APIs, assume that downstream users will need to extend "enumerations" (e.g., result code defined values) as well as structures (e.g., via CONTAINER_OF())
> 
> The above points lead to needing both context and user data parameters. Lets consider when two identical instances exist of communication hardware (e.g., I2C controllers, ethernet controllers, USB OTG controllers, etc). We will create a separate driver instance for each and may create a separate protocol stack instance for each. The protocol component instance data is accessible via `userdata`, the driver component instance is provided through `context`. If the protocol instance is only talking to one driver instance, then `context` may not be needed, but if it is talking to multiple instances of the same component (e.g., timers, GPIOs, etc.), then `context` is used to determine the source of the callback instead of providing a separate callback function for each source.

This is my point in a comment above, the callback from multiple drivers is fine, but the way the helpers for things like a mutex/signal/semaphore/work would not work if there a multiple calling contexts. So why carry that along.

**Comment by gregshue** [comment]
> This is my point in a comment above, the callback from multiple device instances is fine, but the way the helpers for things like a mutex/signal/semaphore/work would not work if there a multiple calling contexts. So why carry that along.

It sounds like the helpers you mentioned don't scale to the many-to-many connections that we need to support (i.e., they don't meet the requirements of an extensible system). What am I missing?

BTW, why would you use a helper for a mutex?  It seems to me that the thread should either block on it singularly or within a condition variable.

**Comment by andyross** [comment]
Driving by again.  I know I promised to avoid the bikeshed, but if it's happening anyway...  No vote from me, trying not to engage too much, just giving opinions:

I gotta be honest: solving the "callback context pointer" vs. "intrusive data structure / CONTAINER_OF" conflict by adding both of them strikes me as the undeniably **worst** resolution possible.  We've all heard of design-by-committee, right?  This is it.

Personally I like intrusive pointers for efficiency reasons (so do the Linux people, FWIW).  But I can live with an opaque context pointer too.  It's just one word of storage for every registerable object.  That's a real and measurable memory cost but not an outrageous one.  And it means the API is clearer and no one needs to explain intrusive data structures on discord, which is a plus.

But putting them back in in addition to the opaque pointer: (1) does nothing to address the efficiency cost, in fact it adds **even more** overhead to pass the needed arguments (in contexts where they can't be inlined out, anyway), and (2) requires that every client of this API then document their own convention about how that context pointer works, which means that we'll be explaining this stuff forever anyway.  I think the only real advantage is just legacy porting overhead given all the pre-existing intrusive code.

Strongly, strongly recommend that if we have people who simply refuse the intrusive scheme, that we go with context pointers only.

Said my bit.  Going to go hide now.

**Comment by gregshue** [comment]
> void the bikeshed, but if it's happening anyway...

I'm trying to avoid that and back us up to what the requirements are on this interface.  Remember, the domain of Zephyr Project includes safety-critical systems, where every line of code must ultimately trace back to at least one top-level system requirement. To minimize future rework and breaking changes, we ought to attempt that now.

For a composable, extensible, reusable event-based platform we need an extensible general API for asynchronously completing a request. 

I propose general asynchronous completion must support:

  d. passing the completion result back across interrupts/threads
  e. the completion result coming from a component other than the one that received the initial request
  f. indicating the request instance being completed
      (which covers multiple outstanding requests from each of multiple instances of a component)
  g. indicating standard alternate completion results due to asynchronous operation (e.g., cancelled)
  h. be extensible to support the propagation of any output data
  i. clearly indicate whether the result is due to a supported condition (e.g., timeout on external data) or an unsupported one (e.g., giving a mutex from a thread that does not own it).  Note: If unsupported conditions are always handled as a catastrophic failure (i.e., k_panic()) then this requirement does not apply.

If we are evaluating this API as an extensible general API for asynchronous completions, then what additional requirements must be met?

If we are **not** evaluating this API as an extensible general API for asynchronous completion, then what are the requirements we are trying to meet with this API?

**Comment by henrikbrixandersen** [comment]
As long as we keep the underlying async APIs based on callbacks on which anybody can build IPC notifications to their own liking, I will not stand in the way of adding this.

I still cannot see how accepting a `context` parameter in the `async.h` callbacks only to throw it away it the `async.c` implementations make for a good, intuitive, and user-friendly API, but I have raised my concerns and I seem to be the minority.

**Comment by gregshue** [comment]
> I still cannot see how accepting a context parameter in the async.h callbacks only to throw it away it the async.c implementations make for a good, intuitive, and user-friendly API, 

Remember that the async.c helpers are there for typical use cases (single instance client, dedicated IPC per outstanding call), but the callback API needs to support the breadth of requirements/use cases. That's why I asked about the requirements.

I would find it more intuitive if we put the callback typedef and common result specification in a separate header file from the helper declarations.

BTW, the helpers are not reusable in the auditable branch as written.  The need to do something in response to the system calls returning an error (e.g., catastrophic failure due to unsupported system behavior). See Rule 17.7 in [Coding Guidelines ](https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html)

**Comment by teburd** [comment]
I've removed all the helpers except async_signal_cb, with updated doc strings to clarify the intended use case is supplanting k_poll_signal in places like spi and i2c async functions where k_poll_signal is currently used. This avoids the conflicting issue over a context pointer, though does sort of leave it hanging in the air.

Updated the commit message to limit the intended scope of the change and what its for.

@henrikbrixandersen to be clear I don't really disagree with what you are saying, I just don't know how to resolve the two, in my mind, conflicting stories. Easily moving from k_poll_signal for things like spi_transcieve_async, and supporting your use case in one go. So maybe this smaller change helps push that off to another PR down the line.

**Comment by teburd** [comment]
Removed the controversial context parameter entirely, documented the need to use CONTAINER_OF for additional context if any is needed, which is compatible with directly using a k_poll_signal

**Review by andyross — APPROVED** [review]
(pleasemergebeforewechangeourmindsagain)

**Review by MaureenHelm — APPROVED** [review]

File: include/zephyr/sys/async.h [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+ *
+ * @brief Utilities supporting asynchronous calls
+ *
+ * @note All callback helpers here are safe to call from an ISR.
+ */
+
+#ifndef ZEPHYR_INCLUDE_SYS_ASYNC_H_
+#define ZEPHYR_INCLUDE_SYS_ASYNC_H_
+
+#include <zephyr/kernel.h>
```
Comment (gmarull) [inline include/zephyr/sys/async.h:18]: nit: this include is not needed here

File: lib/os/Kconfig [hunk 14]
```diff
@@ -125,6 +125,13 @@ config UTF8
 	  Enable the utf8 API. The API implements functions to specifically
 	  handle UTF-8 encoded strings.
 
+config ASYNC
```
Comment (nashif) [inline lib/os/Kconfig:128]: do not like this, it will end up as CONFIG_ASYNC and that can mean a lot of things, the symbol should be more targeted and should provide context in the name.

**Review by nashif — CHANGES_REQUESTED** [review]

**Comment by nordic-krch** [comment]
With lack of `instance` pointer in the callback we are kind of back to where `notify` was. If you want to have that `instance` in the callback (talking about callback case) you need some persistent instance which will be passed as user data and it will hold instance and actual user data. Contrary to `notfiy` where there are tools (maybe a bit convoluted) to handle that (`struct sys_notify`) here responsible is shifted to the user. I don't like it much, I would like to have solution which does not require you to instantiate some object for the most basic case (callback). On the other hand, I'm aware that `void *` is too opaque.

I was thinking about something like this.

In async.h create empty structure:
`struct async_cb_originator { };`

Async callback has pointer to that structure as the first argument:
`async_cb_t(struct async_cb_originator *originator, int result, void *user_data)`

Specific module (or group of modules) provides encapsulation:
```
struct device_cb {
   struct async_cb_originator originator;
   const struct device *dev;
};

static inline const struct device *device_get_instance(struct async_cb_originator *originator)
{
    struct device_cb *data = CONTAINER_OF(originator, struct device_cb, originator);
    return data->dev;
}
```
Specific module implements callback like:
```
void callback(struct device *dev, int result, void *user_data, async_cb_t cb)
{
   struct device_cb data = { .dev = dev };
   cb(&data.origintor, result, user_data);
}
```

User implements own callback as:
```
void user_callback(struct async_originator *originator, int result, void *user_data)
{
   const struct device *dev = device_get_instance(originator);
   ....
}
```

With this approach we have type safe, generic callback without need of instantiating some object and using `CONTAINER_OF` to get originator (device in case of drivers).

To handle signal we would need to create a structure:
```
struct async_poll_signal {
  struct k_poll_signal signal;
  void *originator;
};
```
And module specific conversion function:
```
static inline const struct device *device_get_instance_from_signal(struct async_poll_signal *poll_signal)
{
  return (const struct device *)poll_signal->originator;
}
```

Used like:
```
struct async_poll_signal poll_signal;

spi_transfer(..., async_signal_cb, &poll_signal.signal);

k_poll_signal_check(&poll_signal.signal, ...);
dev = device_get_instance_from_signal(&poll_signal);
```

**Comment by nordic-krch** [comment]
I've implemented my proposal: https://github.com/zephyrproject-rtos/zephyr/commit/d2bb81a63ae728881fdec09ac3ead3fceb0d3894

In the test there is a part that would need to be implemented by the module which supports asynchronous calls using async.h.

I think that using `struct async_callee *` instead of `typedef void *` gives a bit better protection against misusing function for getting callee data.

**Comment by teburd** [comment]
I'm closing this in favor of more directly, from the i2c.h and spi.h interfaces, providing both a callback and a k_poll_signal wrapper of the callback API. This makes more sense...

1. The k_poll_signal wrapper may be, at some point marked as a __syscall while the callback variant will never be a syscall. This is quite nice.
2. No confusion about how to keep call context.
3. Users of the async SPI interface will not need to refactor anything.
